File: expandableRecyclerView/src/main/java/pokercc/android/expandablerecyclerview/ExpandableAdapter.java
package pokercc.android.expandablerecyclerview;

import android.os.Looper;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 可展开的数据适配器
 *
 * @param <Parent>
 * @param <Children>
 * @author pokercc
 * @date 2019-6-2 11:38:13
 */
public abstract class ExpandableAdapter<Parent, Children> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    static final int GROUP_VIEW_TYPE = 1;
    static final int CHILDREN_VIEW_TYPE = 2;
    static final int GROUP_IS_EXPAND_FLAG = 3 << 24;
    static final int GROUP_INDEX_FLAG = 3 << 24 + 1;
    static final int CHILDREN_INDEX_FLAG = 3 << 24 + 2;
    private static final Object GROUP_EXPAND_CHANGE = new Object();

    public static class Group<Parent, Children> {
        public final Parent parent;
        public final List<Children> children;
        public boolean expanded = false;

        public Group(@NonNull Parent parent, @Nullable List<Children> children) {
            this.parent = parent;
            this.children = children == null ? Collections.<Children>emptyList() : children;
        }

        public int getViewHolderCount() {
            if (expanded) {
                return children.size() + 1;
            } else {
                return 1;
            }
        }

    }

    private class RecyclerViewItem {
        final Parent parent;
        final Children children;
        final int viewType;
        final int groupPosition;
        final int childPosition;

        private RecyclerViewItem(Parent parent, int groupPosition) {
            this.parent = parent;
            this.children = null;
            this.viewType = GROUP_VIEW_TYPE;
            this.groupPosition = groupPosition;
            this.childPosition = 0;
        }

        private RecyclerViewItem(Children children, int groupPosition, int childPosition) {
            this.parent = null;
            this.children = children;
            this.viewType = CHILDREN_VIEW_TYPE;
            this.groupPosition = groupPosition;
            this.childPosition = childPosition;
        }
    }

    private final List<Group<Parent, Children>> groups;
    private final List<RecyclerViewItem> recyclerViewItemList = new ArrayList<>();

    private boolean onlyOneGroupExpand = false;

    protected ExpandableAdapter(List<Group<Parent, Children>> groups) {
        this.groups = groups == null ? Collections.<Group<Parent, Children>>emptyList() : groups;
        setDataInternal();
    }

    /**
     * 设置只展开一个group
     *
     * @param onlyOneGroupExpand
     */
    public void setOnlyOneGroupExpand(boolean onlyOneGroupExpand) {
        this.onlyOneGroupExpand = onlyOneGroupExpand;
    }

    private OnChildrenItemClickListener<Children> onChildrenItemClickListener;

    public void setOnChildrenItemClickListener(OnChildrenItemClickListener<Children> onChildrenItemClickListener) {
        this.onChildrenItemClickListener = onChildrenItemClickListener;
    }

    private final ExpandableItemAnimator expandableItemAnimator = new ExpandableItemAnimator(this);

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        recyclerView.setItemAnimator(expandableItemAnimator);
    }

    /**
     * 某个group是否展开
     *
     * @param groupPosition
     * @return
     */
    public boolean isExpand(int groupPosition) {
        return groups.get(groupPosition).expanded;
    }


    private void setDataInternal() {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            throw new IllegalStateException("main on main thread");
        }
        recyclerViewItemList.clear();
        for (int i = 0; i < groups.size(); i++) {
            Group<Parent, Children> group = groups.get(i);
            recyclerViewItemList.add(new RecyclerViewItem(group.parent, i));
            if (group.expanded) {
                for (int j = 0; j < group.children.size(); j++) {
                    recyclerViewItemList.add(new RecyclerViewItem(group.children.get(j), i, j));
                }
            }
        }

    }

    private void performGroupExpandChange(int group, boolean expand) {
        int groupLayoutPosition = getGroupLayoutPosition(group);
        onGroupExpandChange(group, groupLayoutPosition, expand);
        notifyGroupChange(group, GROUP_EXPAND_CHANGE);
    }

    public void onGroupExpandChange(int group, int adapterPosition, boolean expand) {

    }


    /**
     * 通知group布局刷新
     *
     * @param groupPosition
     * @param refreshToken
     */
    public final void notifyGroupChange(int groupPosition, Object refreshToken) {
        notifyItemChanged(getGroupLayoutPosition(groupPosition), refreshToken);
    }

    /**
     * 通知子item局部刷新
     *
     * @param groupPosition
     * @param childrenPosition
     * @param refreshToken
     */
    public final void notifyChildrenChange(int groupPosition, int childrenPosition, Object refreshToken) {
        Group<Parent, Children> item = groups.get(groupPosition);
        if (item.expanded) {
            notifyItemChanged(getChildrenLayoutPosition(groupPosition, childrenPosition), refreshToken);
        }

    }


    /**
     * 展开一个group
     *
     * @param group
     * @param anim
     */
    public final void expandGroup(int group, boolean anim) {
        if (onlyOneGroupExpand) {
            for (int i = 0; i < groups.size(); i++) {
                Group<Parent, Children> item = groups.get(i);
                if (i == group && !item.expanded) {
                    item.expanded = true;
                    // 改变数据源布局刷新
                    setDataInternal();
                    performGroupExpandChange(i, true);
                    if (anim) {
                        notifyItemRangeInserted(getGroupLayoutPosition(group) + 1, item.children.size());
                    } else {
                        notifyDataSetChanged();
                    }
                } else if (item.expanded) {
                    item.expanded = false;
                    // 改变数据源布局刷新
                    setDataInternal();
                    performGroupExpandChange(i, false);
                    if (anim) {
                        notifyItemRangeRemoved(getGroupLayoutPosition(i) + 1, item.children.size());
                    } else {
                        notifyDataSetChanged();
                    }
                }
            }
        } else {
            Group<Parent, Children> item = groups.get(group);
            if (!item.expanded) {
                item.expanded = true;
                // 改变数据源布局刷新
                setDataInternal();
                performGroupExpandChange(group, true);

                if (anim) {
                    notifyItemRangeInserted(getGroupLayoutPosition(group) + 1, item.children.size());
                } else {
                    notifyDataSetChanged();
                }
            }
        }

    }

    /**
     * 折叠一个group
     *
     * @param group
     * @param anim
     */
    public final void collapseGroup(int group, boolean anim) {
        Group<Parent, Children> item = groups.get(group);
        if (item.expanded) {
            item.expanded = false;
            // 改变数据源布局刷新
            setDataInternal();
            if (anim) {
                notifyItemRangeRemoved(getGroupLayoutPosition(group) + 1, item.children.size());
            } else {
                notifyDataSetChanged();
            }
            performGroupExpandChange(group, false);
        }
    }

    private int getGroupLayoutPosition(int groupPosition) {
        int layoutPosition = 0;

        List<Group<Parent, Children>> itemData = this.groups.subList(0, groupPosition);
        for (Group<Parent, Children> item : itemData) {
            layoutPosition += item.getViewHolderCount();
        }
        return layoutPosition;
    }

    private int getChildrenLayoutPosition(int groupPosition, int childrenPosition) {

        return getGroupLayoutPosition(groupPosition) + 1 + childrenPosition;
    }

    @NonNull
    @Override
    public final RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        if (viewType == GROUP_VIEW_TYPE) {
            RecyclerView.ViewHolder parentViewHolder = onCreateParentViewHolder(viewGroup);
            parentViewHolder.itemView.setZ(1);
            return parentViewHolder;
        } else {
            RecyclerView.ViewHolder childrenViewHolder = onCreateChildrenViewHolder(viewGroup);
//            childrenViewHolder.itemView.setZ(100);
            return childrenViewHolder;
        }
    }

    protected abstract RecyclerView.ViewHolder onCreateParentViewHolder(@NonNull ViewGroup viewGroup);

    protected abstract RecyclerView.ViewHolder onCreateChildrenViewHolder(@NonNull ViewGroup viewGroup);

    @Override
    public final int getItemViewType(int position) {
        return recyclerViewItemList.get(position).viewType;
    }

    @Override
    public final void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int i) {
        throw new UnsupportedOperationException("not implementation");
    }

    @Override
    public final void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, final int position, @NonNull List<Object> payloads) {
        final RecyclerViewItem recyclerViewItem = recyclerViewItemList.get(holder.getAdapterPosition());
        holder.itemView.setTag(GROUP_INDEX_FLAG, recyclerViewItem.groupPosition);
        holder.itemView.setTag(CHILDREN_INDEX_FLAG, recyclerViewItem.childPosition);
        if (getItemViewType(holder.getAdapterPosition()) == GROUP_VIEW_TYPE) {
            if (payloads.isEmpty()) {
                final boolean expand = isExpand(recyclerViewItem.groupPosition);
                holder.itemView.setTag(GROUP_IS_EXPAND_FLAG, expand);
                onBindParentViewHolder(holder, recyclerViewItem.parent, recyclerViewItem.groupPosition, expand);
                holder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (isExpand(recyclerViewItem.groupPosition)) {
                            collapseGroup(recyclerViewItem.groupPosition, true);
                        } else {
                            expandGroup(recyclerViewItem.groupPosition, true);
                        }
                    }
                });
            } else {
                for (Object payload : payloads) {
                    if (GROUP_EXPAND_CHANGE == payload) {
                        final boolean expand = isExpand(recyclerViewItem.groupPosition);
                        holder.itemView.setTag(GROUP_IS_EXPAND_FLAG, expand);
                        long animDuration = expand ? expandableItemAnimator.getAddDuration() : expandableItemAnimator.getRemoveDuration();
                        onBindParentViewHolderExpandChange(holder, recyclerViewItem.parent, recyclerViewItem.groupPosition, animDuration, expand);
                    } else {
                        onBindParentViewHolder(holder, recyclerViewItem.parent, recyclerViewItem.groupPosition, payload);
                    }
                }
            }
        } else {
            if (payloads.isEmpty()) {
                onBindChildrenViewHolder(holder, recyclerViewItem.children, recyclerViewItem.groupPosition, recyclerViewItem.childPosition);
                holder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (onChildrenItemClickListener != null) {
                            onChildrenItemClickListener.onItemClick(recyclerViewItem.children, recyclerViewItem.groupPosition, recyclerViewItem.childPosition);
                        }
                    }
                });
            } else {
                for (Object payload : payloads) {
                    onBindChildrenViewHolder(holder, recyclerViewItem.children, recyclerViewItem.groupPosition, recyclerViewItem.childPosition, payload);
                }
            }
        }

    }

    protected abstract void onBindChildrenViewHolder(@NonNull RecyclerView.ViewHolder holder, Children children, int group, int childrenPosition);

    protected void onBindChildrenViewHolder(@NonNull RecyclerView.ViewHolder holder, Children children, int group, int childrenPosition, Object refreshToken) {
        onBindChildrenViewHolder(holder, children, group, childrenPosition);
    }

    protected abstract void onBindParentViewHolder(@NonNull RecyclerView.ViewHolder holder, Parent parent, int groupPosition, boolean expand);

    protected void onBindParentViewHolder(@NonNull RecyclerView.ViewHolder holder, Parent parent, int groupPosition, Object refreshToken) {
        onBindParentViewHolder(holder, parent, groupPosition, isExpand(groupPosition));
    }

    protected abstract void onBindParentViewHolderExpandChange(@NonNull RecyclerView.ViewHolder holder, Parent parent, int groupPosition, long animDuration, boolean expand);


    public interface OnChildrenItemClickListener<T> {
        void onItemClick(T t, int groupPosition, int childrenPosition);

    }

    @Override
    public final int getItemCount() {
        return recyclerViewItemList.size();
    }


    public final int getGroupCount() {
        return groups.size();
    }

    public final int getChildrenCount(int groupIndex) {
        return groups.get(groupIndex).children.size();
    }

    /**
     * 获取组index
     *
     * @param viewHolder
     * @return
     */
    public final int getGroupIndex(RecyclerView.ViewHolder viewHolder) {
        Object index = viewHolder.itemView.getTag(GROUP_INDEX_FLAG);
        if (index instanceof Integer) {
            return (int) index;
        } else {
            return RecyclerView.NO_POSITION;
        }
    }

    /**
     * 获取children index
     *
     * @param viewHolder
     * @return
     */
    public final int getChildrenIndex(RecyclerView.ViewHolder viewHolder) {
        Object index = viewHolder.itemView.getTag(CHILDREN_INDEX_FLAG);
        if (index instanceof Integer) {
            return (int) index;
        } else {
            return RecyclerView.NO_POSITION;
        }
    }
}

File: expandableRecyclerView/src/main/java/pokercc/android/expandablerecyclerview/ExpandableAdapter.kt
@file:Suppress("MemberVisibilityCanBePrivate")

package pokercc.android.expandablerecyclerview

import android.os.Looper
import android.os.Parcel
import android.os.Parcelable
import android.util.Log
import android.util.SparseBooleanArray
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.annotation.UiThread
import androidx.core.util.putAll
import androidx.core.util.set
import androidx.recyclerview.widget.RecyclerView

/**
 * ExpandableAdapter working with ExpandableRecyclerView
 * @author pokercc
 * @date 2019-6-2 11:38:13
 * */
@UiThread
abstract class ExpandableAdapter<VH : ExpandableAdapter.ViewHolder> : RecyclerView.Adapter<VH>() {
    companion object {
        @Suppress("MayBeConstant")
        var DEBUG = BuildConfig.DEBUG
        private const val LOG_TAG = "ExpandableAdapter"
        private val GROUP_EXPAND_CHANGE = Any()
    }

    open class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal lateinit var layoutItemPosition: ItemPosition
    }

    data class ItemPosition(var groupPosition: Int, var childPosition: Int?)

    private val tempItemPosition = ItemPosition(0, null)

    private val expandState = SparseBooleanArray()

    var onlyOneGroupExpand = false

    var enableAnimation = true

    private var recyclerView: RecyclerView? = null

    @CallSuper
    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        require(recyclerView is ExpandableRecyclerView)
        this.recyclerView = recyclerView
    }

    @CallSuper
    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        this.recyclerView = null
    }

    /**
     * Whether the  special group is expand
     *
     * @param groupPosition
     * @return
     */
    fun isExpand(groupPosition: Int): Boolean {
        val groupCount = getGroupCount()
        require(groupPosition in 0 until groupCount) {
            "$groupPosition must in 0 until $groupCount"
        }
        return expandState[groupPosition]
    }

    private fun setExpand(groupPosition: Int, expand: Boolean) {
        expandState[groupPosition] = expand
        onGroupExpandChange(groupPosition, expand)
        notifyGroupChange(groupPosition, GROUP_EXPAND_CHANGE)
    }


    @Deprecated("deprecated", replaceWith = ReplaceWith("onGroupExpandChange(int, boolean)"))
    protected open fun onGroupExpandChange(
        groupPosition: Int,
        adapterPosition: Int,
        expand: Boolean
    ) = Unit

    protected open fun onGroupExpandChange(
        groupPosition: Int,
        expand: Boolean
    ) {
        @Suppress("DEPRECATION")
        onGroupExpandChange(groupPosition, getGroupAdapterPosition(groupPosition), expand)
    }

    fun expandAllGroup() {
        onlyOneGroupExpand = false
        for (i in 0 until getGroupCount()) {
            expandState[i] = true
        }
        notifyDataSetChanged()
    }

    fun collapseAllGroup() {
        for (i in 0 until getGroupCount()) {
            expandState[i] = false
        }
        notifyDataSetChanged()
    }

    /**
     *
     * @param groupPosition
     * @param anim
     */
    fun expandGroup(groupPosition: Int, anim: Boolean) {
        val groupCount = getGroupCount()
        require(groupPosition in 0 until groupCount) {
            "$groupPosition must in 0 until $groupCount"
        }
        if (!onlyOneGroupExpand) {
            if (!isExpand(groupPosition)) {
                setExpand(groupPosition, true)
                if (anim) {
                    val adapterPosition = getChildAdapterPosition2(groupPosition, 0) ?: return
                    notifyItemRangeInserted(adapterPosition, getChildCount(groupPosition))
                } else {
                    notifyDataSetChanged()
                }
            }
            return
        }

        if (anim) {
            for (i in 0 until getGroupCount()) {
                if (i == groupPosition && !isExpand(i)) {
                    setExpand(i, true)
                    val childAdapterPosition = getChildAdapterPosition2(i, 0) ?: 0
                    notifyItemRangeInserted(childAdapterPosition, getChildCount(i))
                } else if (isExpand(i)) {
                    val childAdapterPosition = getChildAdapterPosition2(i, 0) ?: 0
                    setExpand(i, false)
                    notifyItemRangeRemoved(childAdapterPosition, getChildCount(i))
                }
            }
        } else {
            for (i in 0 until getGroupCount()) {
                if (i == groupPosition && !isExpand(i)) {
                    setExpand(i, true)
                } else if (isExpand(i)) {
                    setExpand(i, false)
                }
            }
            notifyDataSetChanged()
        }

    }

    /**
     *
     * @param groupPosition
     * @param anim
     */
    fun collapseGroup(groupPosition: Int, anim: Boolean) {
        val groupCount = getGroupCount()
        require(groupPosition in 0 until groupCount) {
            "$groupPosition must in 0 until $groupCount"
        }
        if (!isExpand(groupPosition)) return
        val childAdapterPosition = getChildAdapterPosition2(groupPosition, 0) ?: return
        setExpand(groupPosition, false)
        if (anim) {
            notifyItemRangeRemoved(childAdapterPosition, getChildCount(groupPosition))
        } else {
            notifyDataSetChanged()
        }
    }


    fun getGroupAdapterPosition(groupPosition: Int): Int {
        val groupCount = getGroupCount()
        require(groupPosition in 0 until groupCount) {
            "$groupPosition must in 0 until $groupCount"
        }
        var position = groupPosition
        for (i in 0 until groupPosition) {
            if (isExpand(i)) {
                position += getChildCount(i)
            }
        }
        return position
    }


    @Deprecated(
        message = "Unclear return value",
        replaceWith = ReplaceWith("getChildAdapterPosition2")
    )
    fun getChildAdapterPosition(groupPosition: Int, childPosition: Int): Int {
        return getChildAdapterPosition2(groupPosition, childPosition) ?: RecyclerView.NO_POSITION
    }

    /**
     * Get special child adapter position,if the group is not expand return null.
     */
    fun getChildAdapterPosition2(groupPosition: Int, childPosition: Int): Int? {
        return if (!isExpand(groupPosition)) {
            null
        } else {
            val childCount = getChildCount(groupPosition)
            require(childPosition in 0 until childCount) {
                "$childPosition must in 0 until $$childCount"
            }
            getGroupAdapterPosition(groupPosition) + 1 + childPosition
        }
    }

    final override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): VH {
        return if (isGroup(viewType)) {
            onCreateGroupViewHolder(viewGroup, viewType)
        } else {
            onCreateChildViewHolder(viewGroup, viewType)
        }
    }

    final override fun getItemViewType(position: Int): Int {
        require(position in 0 until itemCount) {
            "$position must in 0 unit $itemCount"
        }
        val (groupPosition, childPosition) = getItemAdapterPosition(position)
        return if (childPosition == null) {
            getGroupItemViewType(groupPosition)
        } else {
            getChildItemViewType(groupPosition, childPosition)
        }
    }

    open fun getGroupItemViewType(groupPosition: Int): Int = 1

    open fun getChildItemViewType(groupPosition: Int, childPosition: Int): Int = -1

    open fun isGroup(viewType: Int): Boolean = viewType > 0

    final override fun onBindViewHolder(viewHolder: VH, position: Int) = Unit

    final override fun onBindViewHolder(holder: VH, position: Int, payloads: MutableList<Any>) {
        val itemPosition = getItemAdapterPosition(position)
        holder.layoutItemPosition = itemPosition.copy()
        if (DEBUG) {
            Log.d(LOG_TAG, "onBindViewHolder $tempItemPosition")
        }
        val (groupPosition, childPosition) = itemPosition
        if (childPosition == null) {
            performBindParentViewHolder(groupPosition, holder, payloads)
        } else {
            onBindChildViewHolder(holder, groupPosition, childPosition, payloads)
        }
    }

    private fun performBindParentViewHolder(
        groupPosition: Int,
        holder: VH,
        payloads: List<Any>
    ) {
        val expand = isExpand(groupPosition)
        if (payloads.isEmpty()) {
            holder.itemView.setOnClickListener {
                if (isExpand(groupPosition)) {
                    collapseGroup(groupPosition, enableAnimation)
                } else {
                    expandGroup(groupPosition, enableAnimation)
                }
            }
        }
        onBindGroupViewHolder(holder, groupPosition, expand, payloads)
        if (payloads.any { it == GROUP_EXPAND_CHANGE }) {
            val animDuration =
                recyclerView?.itemAnimator?.let { if (expand) it.addDuration else it.removeDuration }
            onGroupViewHolderExpandChange(holder, groupPosition, animDuration ?: 300, expand)
        }
    }

    fun onSaveInstanceState(): Parcelable {
        return ExpandableState(expandState)
    }

    fun onRestoreInstanceState(state: Parcelable?) {
        (state as? ExpandableState)?.expandState?.let {
            expandState.clear()
            expandState.putAll(it)
        }
    }


    /**
     *
     * Notify any registered observers that the group item at <code>groupPosition</code> has changed with
     * an optional payload object.
     *
     * @param groupPosition
     * @param payload
     */
    fun notifyGroupChange(groupPosition: Int, payload: Any? = null) {
        notifyItemChanged(getGroupAdapterPosition(groupPosition), payload)
    }

    /**
     * Notify any registered observers that the child item at <code>groupPosition,childPosition</code> has changed with
     * an optional payload object.
     *
     * @param groupPosition
     * @param childPosition
     * @param payload
     */
    fun notifyChildChange(
        groupPosition: Int,
        childPosition: Int,
        payload: Any? = null
    ) {
        if (isExpand(groupPosition)) {
            val adapterPosition = getChildAdapterPosition2(groupPosition, childPosition) ?: return
            notifyItemChanged(adapterPosition, payload)
        }
    }

    fun notifyGroupInserted(groupPosition: Int) {
        // FIXME: 2020/12/18 Not strict
        notifyItemInserted(getGroupAdapterPosition(groupPosition))
    }

    fun notifyGroupRangeInserted(range: IntRange) {
        // FIXME: 2020/12/18 Not strict
        notifyItemRangeInserted(getGroupAdapterPosition(range.first), range.last - range.first)
    }

    fun notifyGroupRemove(groupPosition: Int) {
        // FIXME: 2020/12/18 Not strict
        notifyItemRemoved(getGroupAdapterPosition(groupPosition))
    }

    fun notifyGroupRangeRemove(range: IntRange) {
        // FIXME: 2020/12/18 Not strict
        notifyItemRangeRemoved(getGroupAdapterPosition(range.first), range.last - range.first)
    }

    fun notifyChildInserted(groupPosition: Int, childPosition: Int) {
        if (isExpand(groupPosition)) {
            val adapterPosition = getChildAdapterPosition2(groupPosition, childPosition) ?: return
            notifyItemInserted(adapterPosition)
        }
    }

    fun notifyChildRangeInserted(groupPosition: Int, range: IntRange) {
        if (isExpand(groupPosition)) {
            val adapterPosition = getChildAdapterPosition2(groupPosition, range.first) ?: return
            notifyItemRangeInserted(adapterPosition, range.last - range.first)
        }
    }

    fun notifyGroupMove(fromGroupPosition: Int, toGroupPosition: Int) {
        // FIXME: 2020/12/18 Not strict
        notifyItemMoved(
            getGroupAdapterPosition(fromGroupPosition),
            getGroupAdapterPosition(toGroupPosition)
        )
    }

    fun notifyChildMove(groupPosition: Int, fromChildPosition: Int, toChildPosition: Int) {
        if (isExpand(groupPosition)) {
            val startPosition = getChildAdapterPosition2(groupPosition, fromChildPosition) ?: return
            val endPosition = getChildAdapterPosition2(groupPosition, toChildPosition) ?: return
            notifyItemMoved(startPosition, endPosition)
        }
    }

    fun notifyChildRemove(groupPosition: Int, childPosition: Int) {
        if (isExpand(groupPosition)) {
            val adapterPosition = getChildAdapterPosition2(groupPosition, childPosition) ?: return
            notifyItemRemoved(adapterPosition)
        }
    }

    fun notifyChildRangeRemove(groupPosition: Int, range: IntRange) {
        if (isExpand(groupPosition)) {
            val adapterPosition = getChildAdapterPosition2(groupPosition, range.first) ?: return
            notifyItemRangeRemoved(adapterPosition, range.last - range.first)
        }
    }

    /**
     * Clear expand state ,working with notifyDataSetChange()
     */
    fun clearExpandState() {
        expandState.clear()
    }

    final override fun getItemCount(): Int {
        var itemCount = 0
        for (g in 0 until getGroupCount()) {
            itemCount++
            if (isExpand(g)) {
                itemCount += getChildCount(g)
            }
        }
        return itemCount
    }


    protected abstract fun onCreateGroupViewHolder(
        viewGroup: ViewGroup, viewType: Int
    ): VH

    protected abstract fun onCreateChildViewHolder(
        viewGroup: ViewGroup,
        viewType: Int
    ): VH


    protected abstract fun onBindChildViewHolder(
        holder: VH,
        groupPosition: Int,
        childPosition: Int,
        payloads: List<Any>
    )

    protected abstract fun onBindGroupViewHolder(
        holder: VH,
        groupPosition: Int,
        expand: Boolean,
        payloads: List<Any>
    )

    protected abstract fun onGroupViewHolderExpandChange(
        holder: VH,
        groupPosition: Int,
        animDuration: Long,
        expand: Boolean
    )

    abstract fun getGroupCount(): Int

    abstract fun getChildCount(groupPosition: Int): Int

    fun getItemLayoutPosition(viewHolder: RecyclerView.ViewHolder): ItemPosition =
        (viewHolder as ViewHolder).layoutItemPosition

    fun getItemAdapterPosition(viewHolder: RecyclerView.ViewHolder): ItemPosition? {
        return if (viewHolder.adapterPosition == RecyclerView.NO_POSITION) {
            null
        } else {
            getItemAdapterPosition(viewHolder.adapterPosition)
        }
    }

    fun getItemAdapterPosition(adapterPosition: Int): ItemPosition {
        require(Looper.myLooper() == Looper.getMainLooper()) {
            "Must run on ui thread"
        }
        require(adapterPosition in 0 until itemCount) {
            "$adapterPosition must in 0 unit $itemCount"
        }
        tempItemPosition.groupPosition = RecyclerView.NO_POSITION
        tempItemPosition.childPosition = null
        var position = -1
        A@ for (g in 0 until getGroupCount()) {
            position++
            if (position == adapterPosition) {
                tempItemPosition.groupPosition = g
                tempItemPosition.childPosition = null
                break@A
            }
            if (!isExpand(g)) continue
            for (c in 0 until getChildCount(g)) {
                position++
                if (position == adapterPosition) {
                    tempItemPosition.groupPosition = g
                    tempItemPosition.childPosition = c
                    break@A
                }
            }
        }
        return tempItemPosition
    }

    /**
     *
     *
     * @param viewHolder
     * @return
     */
    @Deprecated(
        message = "Mix adapterPosition and layoutPosition",
        replaceWith = ReplaceWith("getItemAdapterPosition() or getItemLayoutPosition()")
    )
    fun getChildPosition(viewHolder: RecyclerView.ViewHolder): Int {
        return getItemLayoutPosition(viewHolder).childPosition ?: RecyclerView.NO_POSITION
    }

    /**
     *
     * @param viewHolder
     * @return
     */
    @Deprecated(
        message = "Mix adapterPosition and layoutPosition",
        replaceWith = ReplaceWith("getItemAdapterPosition() or getItemLayoutPosition()")
    )
    fun getGroupPosition(viewHolder: RecyclerView.ViewHolder): Int {
        return getItemLayoutPosition(viewHolder).groupPosition
    }

    class ExpandableState(var expandState: SparseBooleanArray?) : Parcelable {

        constructor(parcel: Parcel) : this(parcel.readSparseBooleanArray())

        override fun writeToParcel(parcel: Parcel, flags: Int) {
            parcel.writeSparseBooleanArray(expandState)
        }

        override fun describeContents(): Int = 0

        companion object CREATOR : Parcelable.Creator<ExpandableState> {
            override fun createFromParcel(parcel: Parcel): ExpandableState {
                return ExpandableState(parcel)
            }

            override fun newArray(size: Int): Array<ExpandableState?> {
                return arrayOfNulls(size)
            }
        }
    }
}

